internal/vm/importers/glob.go

```diff
@@ -0,0 +1,289 @@
+	}
+	if p.stripExtension {
+		pos := strings.LastIndex(name, ".")
+		if pos > 0 {
+			name = name[:pos]
+		}
+	}
+	finalElements := append(dirs, name)
+	return filepath.Join(finalElements...)
+}
+
+func (p globImportParams) String() string {
+	return fmt.Sprintf("?dirs=%d&strip-extension=%t&verb=%s", p.dirs, p.stripExtension, p.verb)
+}
+
+func newGlobParams(query url.Values) (params globImportParams, err error) {
+	for k := range query {
+		if !supportedParams[k] {
+			return params, fmt.Errorf("invalid query parameter '%s', allowed values are: %s", k, supportedParamsDisplay)
+		}
+	}
+	verb := query.Get(globParamVerb)
+	if verb == "" {
+		verb = "import"
+	}
+	if verb != "import" && verb != "importstr" {
+		return params, fmt.Errorf("'%s' parameter for glob import must be one of 'import' or 'importstr', found '%s'", globParamVerb, verb)
+	}
+	dirs := -1
+	dirsStr := query.Get(globParamDirs)
+	if dirsStr != "" {
+		l, err := strconv.Atoi(dirsStr)
+		if err != nil {
+			return params, fmt.Errorf("invalid value '%s' for '%s' parameter, %v", dirsStr, globParamDirs, err)
+		}
+		if l < 0 {
+			l = -1
+		}
+		dirs = l
+	}
+	stripExtension := false
+	stripStr := query.Get(globParamStripExtension)
+	if stripStr != "" {
+		switch stripStr {
+		case "false":
+		case "true":
+			stripExtension = true
+		default:
+			return params, fmt.Errorf("invalid value '%s' for '%s' parameter, must be 'true' or 'false'", stripStr, globParamStripExtension)
+		}
+	}
+	return globImportParams{
+		verb:           verb,
+		dirs:           dirs,
+		stripExtension: stripExtension,
+	}, nil
+}
+
+// globEntry is a cache entry for a glob path resolved from the current working directory.
+type globEntry struct {
+	contents jsonnet.Contents
+	foundAt  string
+	err      error
+}
+
+// GlobImporter provides facilities to import a bag of files using a glob pattern. Note that it will NOT
+// honor any library paths and must be exactly resolved from the caller's location.
+//
+// The input to the glob importer is a URI having a scheme "glob" that has a file pattern in its path
+// that is resolved using Go's glob functionality. The return value is an object that is keyed by file names
+// with values importing the contents of the file.
+//
+// That is, given the following directory structure:
+//
+// 		lib
+//			- a.json
+//			- b.json
+//		caller
+//			- c.libsonnet
+//
+// where c.libsonnet has the following contents
+//
+//		import 'glob:../lib/*.json'
+//
+// evaluating `c.libsonnet` will return jsonnet code of the following form:
+//
+//		{
+//			'../lib/a.json': import '../lib/a.json',
+//			'../lib/b.json': import '../lib/b.json',
+//		}
+//
+// The returned keys of the object and the import mechanism can be customized using query parameters to the glob URI. The parameters supported are:
+//
+// dirs: (number, default=-1) - the number of directory paths to include in the keys of the output. A value of 0 will only have the file name. A negative value preserves the full path.
```

Comment:
What is the intention for this option?

Reply:
this is an ease of use thing to eliminate unnecessary path elements from output keys. It is a number as opposed to a bool (i.e. give me full path or no path at all) because we have cases where need to glob import from multiple sub directories like `teams/*/policy.yaml` and in this `dirs` need to be set to `1` for unambiguous parsing.